### Name: Archit Pandya
#### Assignment: Data Structures


Problem Statement: Analytics Vidhya conducted an assessment test to hire a Data Scientist. The candidates were evaluated on 5 different subjects A, B, C, D, and E. The marks of 5 different subject out of 100 are given below. Help Analytics Vidhyalay to find out the answer following questions:
- Who scored the hightest marks in the subject B?
- What is the average marks scored in the subject C?
- Who scored the hightest perentage of marks?
- If considered only top-4 subjects of a candidate, then who scored the highest percentage of marks?


In [312]:
# Student Marks List
student_marks = [['Name',['A','B','C','D','E']],
                 ['Ankit',[41,34,45,55,63]],
                 ['Aravind',[42,23,34,44,53]],
                 ['Lakshay',[32,23,13,54,67]],
                 ['Gyan',[23,82,23,63,34]],
                 ['Pranav',[21,23,25,56,56]]]

In [325]:
## Extract Name and marks and store into the respective empty list from the given 'student_marks'
## Perform Total and Avg in the loop for the respective Student that is current 'row'
# Store respective subject marks of Student
stud_mark_A = []
stud_mark_B = []
stud_mark_C = []
stud_mark_D = []
stud_mark_E = []

#Store respective student marks of Student
stud1_subject_mark = []
stud2_subject_mark = []
stud3_subject_mark = []
stud4_subject_mark = []
stud5_subject_mark = []

# Store Student Total Marks
stud1_tot_marks = []
stud2_tot_marks = []
stud3_tot_marks = []
stud4_tot_marks = []
stud5_tot_marks = []

# All Students mark
all_stud_tot_marks = []

# Store Subject Total for the student
mark_A = []
mark_B = []
mark_C = []
mark_D = []
mark_E = []

# Loop through the 'student_marks' list and extract Subject marks information for current Students 
for row in range(len(student_marks)):
    if row == 0:
        continue
    # Extract name from the 'student_marks' 
    name = student_marks[row][0]
    # Extract subject marks for the student from the current row  
    marks_A = student_marks[row][1][0]
    marks_B = student_marks[row][1][1]
    marks_C = student_marks[row][1][2]
    marks_D = student_marks[row][1][3]
    marks_E = student_marks[row][1][4]
    
    # sum subject marks for current student and do avgerage
    marks_tot = marks_A + marks_B + marks_C + marks_D + marks_E
    marks_avg = round((marks_tot / 500) * 100,2)
    
    # Append student mark list with Name and Subject Marks
    stud_mark_A.append([marks_A, name])
    stud_mark_B.append([marks_B, name])
    stud_mark_C.append([marks_C, name])
    stud_mark_D.append([marks_D, name])
    stud_mark_E.append([marks_E, name])
    # Append mark list with the respective subject marks
    mark_A.append(marks_A)
    mark_B.append(marks_B)
    mark_C.append(marks_C)
    mark_D.append(marks_D)
    mark_E.append(marks_E)
    
    # Compute total marks for each Student and append into respective student total list
    if row == 1:
        stud1_tot_marks.append([marks_tot, marks_avg, name])
        all_stud_tot_marks.append(stud1_tot_marks)
        # Append individual subject mark by student
        stud1_subject_mark.append([sorted([marks_A, marks_B, marks_C, marks_D, marks_E], reverse = True), name])
    
    elif row == 2:
        stud2_tot_marks.append([marks_tot, marks_avg, name])
        all_stud_tot_marks.append(stud2_tot_marks)
        # Append individual subject mark by student
        stud2_subject_mark.append([sorted([marks_A, marks_B, marks_C, marks_D, marks_E], reverse = True), name])
        
    elif row == 3:
        stud3_tot_marks.append([marks_tot, marks_avg, name])
        all_stud_tot_marks.append(stud3_tot_marks)
        # Append individual subject mark by student
        stud3_subject_mark.append([sorted([marks_A, marks_B, marks_C, marks_D, marks_E], reverse = True), name])
        
    elif row == 4:
        stud4_tot_marks.append([marks_tot, marks_avg, name])
        all_stud_tot_marks.append(stud4_tot_marks)
        # Append individual subject mark by student
        stud4_subject_mark.append([sorted([marks_A, marks_B, marks_C, marks_D, marks_E], reverse = True), name])
        
    elif row == 5:
        stud5_tot_marks.append([marks_tot, marks_avg, name])
        all_stud_tot_marks.append(stud5_tot_marks)
        # Append individual subject mark by student
        stud5_subject_mark.append([sorted([marks_A, marks_B, marks_C, marks_D, marks_E], reverse = True), name])


In [326]:
# All student marks for the respective subjects A,B,C,D,E
print(stud_mark_A)
print(stud_mark_B)
print(stud_mark_C)
print(stud_mark_D)
print(stud_mark_E)

[[41, 'Ankit'], [42, 'Aravind'], [32, 'Lakshay'], [23, 'Gyan'], [21, 'Pranav']]
[[34, 'Ankit'], [23, 'Aravind'], [23, 'Lakshay'], [82, 'Gyan'], [23, 'Pranav']]
[[45, 'Ankit'], [34, 'Aravind'], [13, 'Lakshay'], [23, 'Gyan'], [25, 'Pranav']]
[[55, 'Ankit'], [44, 'Aravind'], [54, 'Lakshay'], [63, 'Gyan'], [56, 'Pranav']]
[[63, 'Ankit'], [53, 'Aravind'], [67, 'Lakshay'], [34, 'Gyan'], [56, 'Pranav']]


In [315]:
# Students total marks
print(stud1_tot_marks)
print(stud2_tot_marks)
print(stud3_tot_marks)
print(stud4_tot_marks)
print(stud5_tot_marks)

[[238, 47.6, 'Ankit']]
[[196, 39.2, 'Aravind']]
[[189, 37.8, 'Lakshay']]
[[225, 45.0, 'Gyan']]
[[181, 36.2, 'Pranav']]


In [316]:
# Total and Avgerage marks by student
print(all_stud_tot_marks)

[[[238, 47.6, 'Ankit']], [[196, 39.2, 'Aravind']], [[189, 37.8, 'Lakshay']], [[225, 45.0, 'Gyan']], [[181, 36.2, 'Pranav']]]


In [317]:
# Students marks in Descening Order
print(stud1_subject_mark)
print(stud2_subject_mark)
print(stud3_subject_mark)
print(stud4_subject_mark)
print(stud5_subject_mark)

[[[63, 55, 45, 41, 34], 'Ankit']]
[[[53, 44, 42, 34, 23], 'Aravind']]
[[[67, 54, 32, 23, 13], 'Lakshay']]
[[[82, 63, 34, 23, 23], 'Gyan']]
[[[56, 56, 25, 23, 21], 'Pranav']]


#### Q1. Who scored the hightest marks in the subject B?
Find the answer by sorting the 'student_mark_B list as descending

In [318]:
sorted(stud_mark_B, reverse=True)

[[82, 'Gyan'], [34, 'Ankit'], [23, 'Pranav'], [23, 'Lakshay'], [23, 'Aravind']]

##### Answer1: Gyan got the higheest marks (82) in Subject B per above

#### Q2. What is the average marks scored in the subject C?
Find the answer by dividing sum of 'mark_C' by length of 'mark_C'


In [319]:
sum(mark_C) / len(mark_C)

28.0

#### Answer2: Average marks scored in the Subject C is 28 per above.

#### Q3. Who scored the hightest perentage of marks?
Find the answer by sorting the 'all_sutd_tot_marks' list as descending

In [320]:
sorted(all_stud_tot_marks,reverse=True)

[[[238, 47.6, 'Ankit']],
 [[225, 45.0, 'Gyan']],
 [[196, 39.2, 'Aravind']],
 [[189, 37.8, 'Lakshay']],
 [[181, 36.2, 'Pranav']]]

##### Answer3: Ankit scored the highest percentage with 47.6%.

#### Q4. If considered only top-4 subjects of a candidate, then who scored the highest percentage of marks?

In [321]:
# Students total marks in descending

print(stud1_subject_mark[0][0])
print(stud2_subject_mark[0][0])
print(stud3_subject_mark[0][0])
print(stud4_subject_mark[0][0])
print(stud5_subject_mark[0][0])

[63, 55, 45, 41, 34]
[53, 44, 42, 34, 23]
[67, 54, 32, 23, 13]
[82, 63, 34, 23, 23]
[56, 56, 25, 23, 21]


In [322]:
#Top 4 subject by student in descending

print(stud1_subject_mark[0][0][0:-1])
print(stud2_subject_mark[0][0][0:-1])
print(stud3_subject_mark[0][0][0:-1])
print(stud4_subject_mark[0][0][0:-1])
print(stud5_subject_mark[0][0][0:-1])

[63, 55, 45, 41]
[53, 44, 42, 34]
[67, 54, 32, 23]
[82, 63, 34, 23]
[56, 56, 25, 23]


In [323]:
# Total for Top 4 subject by student

print(sum(stud1_subject_mark[0][0][0:-1]))
print(sum(stud2_subject_mark[0][0][0:-1]))
print(sum(stud3_subject_mark[0][0][0:-1]))
print(sum(stud4_subject_mark[0][0][0:-1]))
print(sum(stud5_subject_mark[0][0][0:-1]))

204
173
176
202
160


In [340]:
# Average for Top 5 subjects by student

print(sum(stud1_subject_mark[0][0][0:-1])/400*100)
print(sum(stud2_subject_mark[0][0][0:-1])/400*100)
print(sum(stud3_subject_mark[0][0][0:-1])/400*100)
print(sum(stud4_subject_mark[0][0][0:-1])/400*100)
print(sum(stud5_subject_mark[0][0][0:-1])/400*100)

51.0
43.25
44.0
50.5
40.0


##### Answer4: Ankit scored the highest percentage with 51% for Top 4 Subjects